diff --git a/index.d.ts b/index.d.ts
index ac6b9f5f95d258493e5ce026386be0ab88e423c0..a26edf516e8e6f02deff1585eb73b551c7e95cad 100644
--- a/index.d.ts
+++ b/index.d.ts
@@ -3,8 +3,11 @@ type HeaderValue = string | string[] | number | boolean;
 
 type AxiosHeaders = Record<string, HeaderValue | Record<Method & CommonHeaders, HeaderValue>>;
 
+type LowercaseMethod =
+  | 'get' | 'delete' | 'head' | 'options' | 'post'
+  | 'put' | 'patch' | 'purge' | 'link' | 'unlink';
 type MethodsHeaders = {
-  [Key in Method as Lowercase<Key>]: AxiosHeaders;
+  [Key in LowercaseMethod]?: AxiosHeaders;
 };
 
 interface CommonHeaders  {
